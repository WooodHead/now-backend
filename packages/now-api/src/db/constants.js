/* eslint-disable import/prefer-default-export  */

export const TABLES = {
  ACTIVITY: 'now_template',
  DEVICE: 'now_device',
  EVENT: 'now_event',
  RSVP: 'now_rsvp',
  MESSAGE: 'now_messages',
  USER: 'now_user',
};
