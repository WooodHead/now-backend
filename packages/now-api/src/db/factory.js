/* eslint-disable import/no-extraneous-dependencies  */
import { Factory } from 'rosie';
import uuid from 'uuid/v4';
import { address, name, date, internet, random, lorem, hacker } from 'faker';
import randomEmoji from 'random-emoji';
import { LocalDate, ZoneId, LocalDateTime } from 'js-joda';
import {
  EVENT_INVITE_TYPE,
  APP_INVITE_TYPE,
} from '../schema/resolvers/Invitation';

const md5 = () =>
  random
    .hexaDecimal(32)
    .toLowerCase()
    .substring(2);

Factory.define('user')
  .attr('id', uuid)
  .attr('auth0Id', () => random.alphaNumeric(24))
  .attr('email', internet.email)
  .attr('firstName', name.firstName)
  .attr('lastName', name.lastName)
  .attr('bio', lorem.paragraph)
  .attr('location', () => `${address.city()}, ${address.stateAbbr()}`)
  .attr('preferences', {})
  .attr('birthday', () =>
    LocalDate.now()
      .minusYears(20)
      .toString()
  )
  .attr('photoId', md5)
  .attr(
    'photoPreview',
    'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACgAAAAoCAYAAACM/rhtAAAAAXNSR0IArs4c6QAAAmRJREFUWAntmN1Kw0AQhU/8Fwv1F0FEKCgV9UbvvKkPIFJ8AB/C1/BZfAdF9EpBFEUEBQUp0l6opRVEi/Y4LIlx006aWHuRgTHpZHbmy9lssuh81g0dbF0dzPaNlgBGnaFEwUTBqAqY8e/vwM4OUCiYiO7Ytmfw+Bi4vRXXoUlW2wD396VhqRQGD2gL4N0dcH/fwYBGPSIWiwKq/atSsFbTlvudV60CJyduPOwU97hD7WcfH8D2NpBOA0tLwOIikM0CfX32fH/06AjgCjb29ga8vEg9E2t0bArYU8/gfod3vrcnztjcnMASeGoquMXBwe9rrMUb1lhTQBYZGwMeH91yVPXqSnx3FxgZEVgqPD8PDA5K7uWl3Jg7Us4IODvrj9p/twToL/X0BBweind3A5mMgJ6e+jPld5iFogIcH7c3skW5oG5uxG3XGQuzUFSrOAxgEJQ3Hjsgn8E4LcwU/4uCr68A348aUwHGrSDBtCqqAFMpYGBAc7/6HO1zqAJk27hVjB0waCVPTAC5HDA6qlePmdopVr0HWTBIwa0t+TYz5+EBODsTN9srxm2mVVANaFNwYcGFI8T0tPj6OvD8LKDn58D19c8NA3P/HNBxgM1NtrLb8DCwtibOHQy/y1T34gKoVIByWaB7e+3jTVStoH+KV1aAmRlTpvGxvx9gPp07I34KqWAzOFZVA3qnuKu+9vP5xlBBV6k8t2p0jalfM3wPDg1JydVVYHJSUz56jhqQragip2VjI3pjbQX1FLMgd8/Ly7JB1TaImuck/8CMKGGoZzBir5aGJ4AtyeYZlCjoEaOl0y+Le57bWRVobwAAAABJRU5ErkJggg=='
  );

Factory.define('device')
  .option('user', () => ({ id: uuid() }))
  .attr('token', uuid)
  .attr('type', () => random.arrayElement(['ios', 'android']))
  .attr('model', random.randomWord)
  .attr('userId', ['user'], ({ id }) => id);

Factory.define('rsvp')
  .option('user', () => ({ id: uuid() }))
  .option('event', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('action', 'add')
  .attr('userId', ['user'], ({ id }) => id)
  .attr('eventId', ['event'], ({ id }) => id);

Factory.define('appInvite')
  .option('inviter', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('type', APP_INVITE_TYPE)
  .attr('code', () => String(random.number({ max: 999999, min: 100000 })))
  .attr('expiresAt', () =>
    LocalDateTime.now()
      .plusDays(1)
      .toString()
  )
  .attr('inviterId', ['inviter'], ({ id }) => id);

Factory.define('eventInvite')
  .option('inviter', () => ({ id: uuid() }))
  .option('event', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('type', EVENT_INVITE_TYPE)
  .attr('code', () => String(random.number({ max: 999999, min: 100000 })))
  .attr('expiresAt', () =>
    LocalDateTime.now()
      .plusDays(1)
      .toString()
  )
  .attr('inviterId', ['inviter'], ({ id }) => id)
  .attr('eventId', ['event'], ({ id }) => id);

Factory.define('activity')
  .attr('id', uuid)
  .attr('activityDate', () =>
    date
      .soon(1)
      .toISOString()
      .substring(0, 10)
  )
  .attr('title', () => `${hacker.ingverb()} for ${hacker.noun()}`)
  .attr('emoji', () => randomEmoji.random({ count: 1 })[0].character)
  .attr('description', hacker.phrase);

Factory.define('todayActivity')
  .extend('activity')
  .attr('activityDate', () => new Date().toISOString().substring(0, 10));

Factory.define('event')
  .option('activity', () => ({
    id: uuid(),
    activityDate: date
      .soon(1)
      .toISOString()
      .substring(0, 10),
  }))
  .option('location', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('activityId', ['activity'], ({ id }) => id)
  .attr('locationId', ['location'], ({ id }) => id)
  .attr('limit', () => random.number({ max: 20, min: 5 }))
  .attr('time', ['activity'], ({ activityDate: activityDateString }) => {
    const activityDate = LocalDate.parse(activityDateString);
    return date
      .between(
        activityDate.atStartOfDay(),
        activityDate.atTime4(23, 59, 59, 999)
      )
      .toISOString();
  })
  .attr('timezone', () => random.arrayElement(ZoneId.getAvailableZoneIds()));

Factory.define('location')
  .attr('id', uuid)
  .attr('name', hacker.ingverb)
  .attr('address', address.streetAddress)
  .attr('city', address.city)
  .attr('state', address.stateAbbr)
  .attr('postalCode', address.zipCode)
  .attr('country', address.countryCode)
  .attr('neighborhood', hacker.noun)
  .attr('foursquareVenueId', md5)
  .attr(
    'location',
    () => `SRID=4326;POINT(${address.longitude()} ${address.latitude()})`
  );

Factory.define('message')
  .option('event', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('ts', () => date.recent(1).getTime())
  .attr('text', hacker.phrase)
  .attr('eventId', ['event'], ({ id }) => id)
  .attr('userId', ['user'], ({ id }) => id);

export default Factory;
