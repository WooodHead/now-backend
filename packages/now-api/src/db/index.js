import AWS from 'aws-sdk';
import promisify from 'util.promisify';
import { concatMapOfArrays, isDev, promiseDelay } from '../util';

const dynamoOpts = {};
if (isDev()) {
  dynamoOpts.logger = console;
}

const dynamoDb = new AWS.DynamoDB.DocumentClient(dynamoOpts);

const pGet = promisify(dynamoDb.get.bind(dynamoDb));
const pScan = promisify(dynamoDb.scan.bind(dynamoDb));
const pUpdate = promisify(dynamoDb.update.bind(dynamoDb));
const pPut = promisify(dynamoDb.put.bind(dynamoDb));
const pQuery = promisify(dynamoDb.query.bind(dynamoDb));
const pBatchGet = promisify(dynamoDb.batchGet.bind(dynamoDb));

export const get = (table, key = {}) =>
  pGet({
    TableName: table,
    Key: key,
  }).then(response => response.Item);

export const put = (table, item, condition = undefined) =>
  pPut({
    TableName: table,
    Item: item,
    ConditionExpression: condition,
  });

export const update = (
  table,
  key,
  expr,
  values,
  names = undefined,
  condition = undefined
) =>
  pUpdate({
    TableName: table,
    Key: key,
    ExpressionAttributeValues: values,
    ExpressionAttributeNames: names,
    UpdateExpression: expr,
    ConditionExpression: condition,
    ReturnValues: 'ALL_NEW',
  });

export const updateDynamic = (table, key, values, condition = undefined) => {
  const expressions = [];
  const expressionNames = {};
  const expressionValues = {};
  Object.entries(values).forEach(([k, v], i) => {
    const attrName = `#f${i}`;
    const fieldName = `:f${i}`;
    expressions.push(`${attrName}=${fieldName}`);
    expressionNames[attrName] = k;
    expressionValues[fieldName] = v;
  });

  return update(
    table,
    key,
    `set ${expressions.join(', ')}`,
    expressionValues,
    expressionNames,
    condition
  );
};

export const scan = (table, filter = undefined) =>
  pScan({
    TableName: table,
    FilterExpression: filter,
  }).then(response => response.Items);

export const queryRaw = params => pQuery(params);
export const query = params => pQuery(params).then(response => response.Items);

const batchGetHelper = (params, delay = 50, accum = {}) =>
  pBatchGet(params).then(({ Responses, UnprocessedKeys }) => {
    const nextAccum = concatMapOfArrays(accum, Responses);
    if (!UnprocessedKeys || Object.keys(UnprocessedKeys).length === 0) {
      return nextAccum;
    }
    if (delay >= 2000) {
      return Promise.reject(new Error('too many retries'));
    }
    return promiseDelay(delay).then(() =>
      batchGetHelper(
        { ...params, RequestItems: UnprocessedKeys },
        delay * 2,
        nextAccum
      )
    );
  });

export const batchGet = (tableName, keys, keyName = 'id', opts = {}) => {
  const RequestItems = {
    [tableName]: { ...opts, Keys: keys.map(key => ({ [keyName]: key })) },
  };
  return batchGetHelper({ RequestItems }).then(response => response[tableName]);
};

export const createSet = someArray => dynamoDb.createSet(someArray);
