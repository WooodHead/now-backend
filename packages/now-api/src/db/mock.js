/* eslint-disable import/no-extraneous-dependencies  */
import ApolloClient from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';

import schema from '../schema';
import loaders from './loaders';

export const USER_ID = 'aea68a98-591c-11e8-9cdb-872a15ebfd30';
export const mockPromise = (resolver, rejecter) =>
  rejecter ? Promise.reject(rejecter) : Promise.resolve(resolver);

export const mocks = {
  loadMember: jest.fn(),
};

const apolloCache = new InMemoryCache();

let isAdmin = false;

export const setAdmin = b => {
  isAdmin = b;
};

export const client = new ApolloClient({
  cache: apolloCache,
  link: new SchemaLink({
    schema,
    context: () => ({
      token: null,
      user: { id: USER_ID },
      loaders: loaders({ currentUserId: USER_ID }),
      scopes: isAdmin ? ['admin'] : [],
    }),
  }),
});

export const newUserClient = (currentUserAuth0Id, currentUserId) =>
  new ApolloClient({
    cache: apolloCache,
    link: new SchemaLink({
      schema,
      context: {
        token: null,
        currentUserAuth0Id,
        loaders: loaders({ currentUserId }),
      },
    }),
  });
