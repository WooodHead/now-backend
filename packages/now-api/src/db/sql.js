import knex from 'knex';
import knexConfig from '../../knexfile';

const sql = knex(knexConfig[process.env.NODE_ENV]);

export default sql;
