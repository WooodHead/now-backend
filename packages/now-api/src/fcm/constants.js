/* eslint-disable import/prefer-default-export */
export const NEW_EVENTS_TOPIC = 'new-events';
