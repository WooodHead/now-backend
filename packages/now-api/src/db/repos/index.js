import sql from '../sql';
import { SQL_TABLES } from '../constants';
import { putInOrder } from '../../util';

const buildRepo = table => {
  const base = {
    get: (query = {}) =>
      sql(table)
        .where(query)
        .first(),
    all: (query = {}) => sql(table).where(query),
    insert: obj => sql(table).insert(obj),
    update: ({ id, ...otherFields }) =>
      sql(table)
        .where({ id })
        .update(otherFields),
    batch: ids =>
      sql(table)
        .whereIn('id', ids)
        .then(batch => putInOrder(batch, ids)),
    delete: query =>
      sql(table)
        .where(query)
        .delete(),
  };

  base.byId = id => base.get({ id });

  return base;
};

export const User = buildRepo(SQL_TABLES.USERS);
export const Device = buildRepo(SQL_TABLES.DEVICES);
export const Activity = buildRepo(SQL_TABLES.ACTIVITIES);
export const Event = buildRepo(SQL_TABLES.EVENTS);
export const Location = buildRepo(SQL_TABLES.LOCATIONS);
export const Rsvp = buildRepo(SQL_TABLES.RSVPS);
export const Message = buildRepo(SQL_TABLES.MESSAGES);
