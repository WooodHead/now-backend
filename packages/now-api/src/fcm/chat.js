import { messaging } from './client';
import { getTokensForEvent } from './tokens';
import { ellipsize } from '../util';
import { Activity, Event, User } from '../db/repos';

const getEventAndActivity = eventId =>
  Event.byId(eventId).then(event =>
    Activity.byId(event.activityId).then(activity => ({
      event,
      activity,
    }))
  );

const sendChatNotif = ({ eventId, userId, text }) =>
  Promise.all([
    getTokensForEvent(eventId, 'messagesNotification', [userId]),
    getEventAndActivity(eventId),
    User.byId(userId),
  ])
    .then(
      ([
        tokens,
        { activity: { title: activityTitle, emoji } },
        { firstName },
      ]) => {
        if (tokens.length === 0) {
          return Promise.resolve();
        }
        return messaging.sendToDevice(tokens, {
          notification: {
            body: `${emoji} ${ellipsize(
              activityTitle,
              20
            )} @${firstName}: ${text}`,
          },
        });
      }
    )
    .catch(console.warn);

export default sendChatNotif;
