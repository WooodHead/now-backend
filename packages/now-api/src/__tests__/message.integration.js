import gql from 'graphql-tag';
import uuid from 'uuid/v4';

import { client, USER_ID, subClient } from '../db/mock';
import factory from '../db/factory';
import { SQL_TABLES } from '../db/constants';
import sql from '../db/sql';
import chat from '../fcm/chat';
import { Message, Rsvp } from '../db/repos';

jest.mock('../fcm/chat', () => jest.fn());

const event = factory.build('event', {
  id: '516f21b1-aff8-4ae0-be1f-e99b415f3652',
});

const user1 = factory.build('user', {
  id: USER_ID,
});

const user2 = factory.build('user', {
  id: 'e5380f3c-57c8-11e8-b397-7be415c41792',
});

const rsvp1 = factory.build('rsvp', {}, { event, user: user1 });

const rsvp2 = factory.build('rsvp', {}, { event, user: user2 });

const message1 = {
  id: uuid(),
  eventId: event.id,
  text: 'message text 1',
  ts: 1519765415755,
  userId: user1.id,
};

const message2 = {
  id: uuid(),
  eventId: event.id,
  text: 'message text 2',
  ts: 1519765415765,
  userId: user2.id,
};

const message3 = {
  id: uuid(),
  eventId: event.id,
  text: 'message text 3',
  ts: 1519765415775,
  userId: user2.id,
};

const message4 = {
  id: uuid(),
  eventId: event.id,
  text: 'message text 4',
  ts: 1519765415785,
  userId: user2.id,
};

const message5 = {
  id: uuid(),
  eventId: event.id,
  text: 'message text 5',
  ts: 1519765415795,
  userId: user2.id,
};

const truncateTables = () =>
  Promise.all([
    sql(SQL_TABLES.EVENTS).truncate(),
    sql(SQL_TABLES.USERS).truncate(),
    sql(SQL_TABLES.MESSAGES).truncate(),
    sql(SQL_TABLES.RSVPS).truncate(),
  ]);

describe('message', () => {
  beforeEach(() =>
    truncateTables().then(() =>
      Promise.all([
        sql(SQL_TABLES.EVENTS).insert(event),
        sql(SQL_TABLES.USERS).insert([user1, user2]),
        sql(SQL_TABLES.RSVPS).insert([rsvp1, rsvp2]),
        sql(SQL_TABLES.MESSAGES).insert([
          message1,
          message2,
          message3,
          message4,
          message5,
        ]),
      ])
    ));
  afterEach(() => truncateTables());
  it('returns eventMessages', async () => {
    const results = client.query({
      query: gql`
        query getEvent($id: ID!) {
          event(id: $id) {
            messages {
              edges {
                node {
                  event {
                    id
                  }
                  text
                  ts
                  user {
                    id
                  }
                }
              }
            }
          }
        }
      `,
      variables: { id: event.id },
    });
    const { data } = await results;
    expect(data).toMatchSnapshot();
  });
  it('returns first n results', async () => {
    const results = client.query({
      query: gql`
        query getEvent($id: ID!) {
          event(id: $id) {
            messages(first: 2) {
              edges {
                node {
                  event {
                    id
                  }
                  text
                  ts
                  user {
                    id
                  }
                }
              }
            }
          }
        }
      `,
      variables: { id: event.id },
    });
    const { data } = await results;
    expect(data).toMatchSnapshot();
  });
  it('returns last n results', async () => {
    const results = client.query({
      query: gql`
        query getEvent($id: ID!) {
          event(id: $id) {
            messages(last: 2) {
              edges {
                node {
                  event {
                    id
                  }
                  text
                  ts
                  user {
                    id
                  }
                }
              }
            }
          }
        }
      `,
      variables: { id: event.id },
    });
    const { data } = await results;
    expect(data).toMatchSnapshot();
  });
  describe('createMessage', () => {
    it('send message', async done => {
      const sub = subClient();
      sub
        .subscribe({
          query: gql`
            subscription {
              unreadMessagesCount
            }
          `,
        })
        .subscribe({
          next: ({ data }) => {
            // the 5 messages inserted at top, + the one inserted below
            expect(data.unreadMessagesCount).toEqual(6);
            done();
          },
        });

      const results = await sub.mutate({
        mutation: gql`
          mutation message($input: CreateMessageInput!) {
            createMessage(input: $input) {
              edge {
                node {
                  id
                  text
                  ts
                }
              }
            }
          }
        `,
        variables: { input: { eventId: event.id, text: 'new message' } },
      });
      const { data } = results;
      expect(data).toMatchObject({
        createMessage: {
          __typename: 'CreateMessagePayload',
          edge: {
            __typename: 'EventMessagesEdge',
            node: {
              __typename: 'Message',
              id: expect.stringMatching(
                /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i
              ),
            },
          },
        },
      });

      const { id, ts } = data.createMessage.edge.node;
      expect(chat).toBeCalledWith(
        expect.objectContaining({
          eventId: event.id,
          id,
          text: 'new message',
          ts: Number(ts),
          userId: USER_ID,
        })
      );

      const dbMessage = await Message.byId(id);
      expect(dbMessage).toMatchObject({
        eventId: event.id,
        id,
        text: 'new message',
        ts,
        userId: USER_ID,
      });
    });

    it('requires rsvp', async () => {
      await Rsvp.delete({ id: rsvp1.id });
      const results = client.mutate({
        mutation: gql`
          mutation message($input: CreateMessageInput!) {
            createMessage(input: $input) {
              edge {
                node {
                  id
                  text
                  ts
                }
              }
            }
          }
        `,
        variables: { input: { eventId: event.id, text: 'new message' } },
      });

      expect.assertions(1);
      return expect(results).rejects.toEqual(
        new Error(
          'GraphQL error: You must have RSVPed before you can post a message.'
        )
      );
    });
  });
});
