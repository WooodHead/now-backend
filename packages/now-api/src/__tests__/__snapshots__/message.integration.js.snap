// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`message returns eventMessages 1`] = `
Object {
  "event": Object {
    "__typename": "Event",
    "messages": Object {
      "__typename": "EventMessagesConnection",
      "edges": Array [
        Object {
          "__typename": "EventMessagesEdge",
          "node": Object {
            "__typename": "Message",
            "event": Object {
              "__typename": "Event",
              "id": "516f21b1-aff8-4ae0-be1f-e99b415f3652",
              Symbol(id): "Event:516f21b1-aff8-4ae0-be1f-e99b415f3652",
            },
            "text": "message text 5",
            "ts": "1519765415795",
            "user": Object {
              "__typename": "User",
              "id": "e5380f3c-57c8-11e8-b397-7be415c41792",
              Symbol(id): "User:e5380f3c-57c8-11e8-b397-7be415c41792",
            },
            Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"}).messages.edges.0.node",
          },
          Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"}).messages.edges.0",
        },
        Object {
          "__typename": "EventMessagesEdge",
          "node": Object {
            "__typename": "Message",
            "event": Object {
              "__typename": "Event",
              "id": "516f21b1-aff8-4ae0-be1f-e99b415f3652",
              Symbol(id): "Event:516f21b1-aff8-4ae0-be1f-e99b415f3652",
            },
            "text": "message text 4",
            "ts": "1519765415785",
            "user": Object {
              "__typename": "User",
              "id": "e5380f3c-57c8-11e8-b397-7be415c41792",
              Symbol(id): "User:e5380f3c-57c8-11e8-b397-7be415c41792",
            },
            Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"}).messages.edges.1.node",
          },
          Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"}).messages.edges.1",
        },
        Object {
          "__typename": "EventMessagesEdge",
          "node": Object {
            "__typename": "Message",
            "event": Object {
              "__typename": "Event",
              "id": "516f21b1-aff8-4ae0-be1f-e99b415f3652",
              Symbol(id): "Event:516f21b1-aff8-4ae0-be1f-e99b415f3652",
            },
            "text": "message text 3",
            "ts": "1519765415775",
            "user": Object {
              "__typename": "User",
              "id": "e5380f3c-57c8-11e8-b397-7be415c41792",
              Symbol(id): "User:e5380f3c-57c8-11e8-b397-7be415c41792",
            },
            Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"}).messages.edges.2.node",
          },
          Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"}).messages.edges.2",
        },
        Object {
          "__typename": "EventMessagesEdge",
          "node": Object {
            "__typename": "Message",
            "event": Object {
              "__typename": "Event",
              "id": "516f21b1-aff8-4ae0-be1f-e99b415f3652",
              Symbol(id): "Event:516f21b1-aff8-4ae0-be1f-e99b415f3652",
            },
            "text": "message text 2",
            "ts": "1519765415765",
            "user": Object {
              "__typename": "User",
              "id": "e5380f3c-57c8-11e8-b397-7be415c41792",
              Symbol(id): "User:e5380f3c-57c8-11e8-b397-7be415c41792",
            },
            Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"}).messages.edges.3.node",
          },
          Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"}).messages.edges.3",
        },
        Object {
          "__typename": "EventMessagesEdge",
          "node": Object {
            "__typename": "Message",
            "event": Object {
              "__typename": "Event",
              "id": "516f21b1-aff8-4ae0-be1f-e99b415f3652",
              Symbol(id): "Event:516f21b1-aff8-4ae0-be1f-e99b415f3652",
            },
            "text": "message text 1",
            "ts": "1519765415755",
            "user": Object {
              "__typename": "User",
              "id": "aea68a98-591c-11e8-9cdb-872a15ebfd30",
              Symbol(id): "User:aea68a98-591c-11e8-9cdb-872a15ebfd30",
            },
            Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"}).messages.edges.4.node",
          },
          Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"}).messages.edges.4",
        },
      ],
      Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"}).messages",
    },
    Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"})",
  },
  Symbol(id): "ROOT_QUERY",
}
`;

exports[`message returns first n results 1`] = `
Object {
  "event": Object {
    "__typename": "Event",
    "messages": Object {
      "__typename": "EventMessagesConnection",
      "edges": Array [
        Object {
          "__typename": "EventMessagesEdge",
          "node": Object {
            "__typename": "Message",
            "event": Object {
              "__typename": "Event",
              "id": "516f21b1-aff8-4ae0-be1f-e99b415f3652",
              Symbol(id): "Event:516f21b1-aff8-4ae0-be1f-e99b415f3652",
            },
            "text": "message text 5",
            "ts": "1519765415795",
            "user": Object {
              "__typename": "User",
              "id": "e5380f3c-57c8-11e8-b397-7be415c41792",
              Symbol(id): "User:e5380f3c-57c8-11e8-b397-7be415c41792",
            },
            Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"}).messages({\\"first\\":2}).edges.0.node",
          },
          Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"}).messages({\\"first\\":2}).edges.0",
        },
        Object {
          "__typename": "EventMessagesEdge",
          "node": Object {
            "__typename": "Message",
            "event": Object {
              "__typename": "Event",
              "id": "516f21b1-aff8-4ae0-be1f-e99b415f3652",
              Symbol(id): "Event:516f21b1-aff8-4ae0-be1f-e99b415f3652",
            },
            "text": "message text 4",
            "ts": "1519765415785",
            "user": Object {
              "__typename": "User",
              "id": "e5380f3c-57c8-11e8-b397-7be415c41792",
              Symbol(id): "User:e5380f3c-57c8-11e8-b397-7be415c41792",
            },
            Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"}).messages({\\"first\\":2}).edges.1.node",
          },
          Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"}).messages({\\"first\\":2}).edges.1",
        },
      ],
      Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"}).messages({\\"first\\":2})",
    },
    Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"})",
  },
  Symbol(id): "ROOT_QUERY",
}
`;

exports[`message returns last n results 1`] = `
Object {
  "event": Object {
    "__typename": "Event",
    "messages": Object {
      "__typename": "EventMessagesConnection",
      "edges": Array [
        Object {
          "__typename": "EventMessagesEdge",
          "node": Object {
            "__typename": "Message",
            "event": Object {
              "__typename": "Event",
              "id": "516f21b1-aff8-4ae0-be1f-e99b415f3652",
              Symbol(id): "Event:516f21b1-aff8-4ae0-be1f-e99b415f3652",
            },
            "text": "message text 2",
            "ts": "1519765415765",
            "user": Object {
              "__typename": "User",
              "id": "e5380f3c-57c8-11e8-b397-7be415c41792",
              Symbol(id): "User:e5380f3c-57c8-11e8-b397-7be415c41792",
            },
            Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"}).messages({\\"last\\":2}).edges.0.node",
          },
          Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"}).messages({\\"last\\":2}).edges.0",
        },
        Object {
          "__typename": "EventMessagesEdge",
          "node": Object {
            "__typename": "Message",
            "event": Object {
              "__typename": "Event",
              "id": "516f21b1-aff8-4ae0-be1f-e99b415f3652",
              Symbol(id): "Event:516f21b1-aff8-4ae0-be1f-e99b415f3652",
            },
            "text": "message text 1",
            "ts": "1519765415755",
            "user": Object {
              "__typename": "User",
              "id": "aea68a98-591c-11e8-9cdb-872a15ebfd30",
              Symbol(id): "User:aea68a98-591c-11e8-9cdb-872a15ebfd30",
            },
            Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"}).messages({\\"last\\":2}).edges.1.node",
          },
          Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"}).messages({\\"last\\":2}).edges.1",
        },
      ],
      Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"}).messages({\\"last\\":2})",
    },
    Symbol(id): "$ROOT_QUERY.event({\\"id\\":\\"516f21b1-aff8-4ae0-be1f-e99b415f3652\\"})",
  },
  Symbol(id): "ROOT_QUERY",
}
`;
